### DISCUSSION TITLE
unbound breakpoint in browser code when "Launch Browser Frontend" in VSCODE

### DISCUSSION THREAD

**Discussion by stat0s2p** [eclipse-theia/theia#8494] [General] [answered]
# Environment
os version:macox 10.15.6
vscode version:1.49.0
google chrome version:85.0.4183.102
theia version：v1.5.0
Debugger for chrome extension:4.12.10
# Description
![image](https://user-images.githubusercontent.com/24647459/92851502-3693ad00-f420-11ea-8fbf-9dc659b62d26.png)

i have run ”Launch Browser Frontend“.after it open new chrome browser,i add a breakpoint on `examples/api-samples/src/browser/menu/sample-menu-contribution.ts` L34 but it is grey and show `unbound breakpoint`.

can anyone help me debug browser code in VSCODE?

**Comment by vince-fugnitto** [comment]
@stat0s2p thank you for the discussion, have you viewed the documentation regarding [debugging](https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#debugging)?

**Reply by stat0s2p** [reply]
Yes.I build and debug my project follow these steps:https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#debug-the-browser-examples-frontend-and-backend-at-the-same-time.

the backend nodejs code breakpoint in vscode work oky but chrome breakpoint not work.
I found ts file in chrome devtools source panel then set breakpoint,it not work too.

Besides,I install nodejs12 and python in conda virtual environment instead of install them globally to avoid conflicts of other projects.should these has some effects?

**Reply by vince-fugnitto** [reply]
> Besides,I install nodejs12 and python in conda virtual environment instead of install them globally to avoid conflicts of other projects.should these has some effects?

I do not think so no.

I verified myself and I am able to successfully debug both the `browser frontend` and `browser backend` simultaneously using the steps mentioned here: https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md#debug-the-browser-examples-frontend-and-backend-at-the-same-time.

Note that I am using vscode to launch both configurations when using `theia` as a workspace.

**Reply by stat0s2p** [reply]
ok,maybe some problem with my environment.I will install a new system and test it again in working days.Thank you very much.

**Comment by stat0s2p** [comment, marked as answer]
@vince-fugnitto 
Dear,I installed a fresh ubuntu desktop os in vmware and follow every steps in https://github.com/eclipse-theia/theia/blob/master/doc/Developing.md but it is still has gray break point in vscode.

then I add a breakpoint in chrome dev tools,I found it worked and active the vscode source file,but if I add a breakpoint direct in vscode source,it is still gray.
![Snipaste_2020-09-12_03-54-05](https://user-images.githubusercontent.com/24647459/92967358-b024ad00-f4ab-11ea-9ffe-89a15c7fdd8e.png)

also,If I add break point in `examples/api-samples/src/browser/menu/sample-menu-contribution.ts`,it canot active the breakpoint both in chrome devtool and vscode source file.

althrough I can set breakpoint in devtool,but is too hard to location the code in devtool.Is a way that I can set a break point in vscode directly?

**Reply by vince-fugnitto** [reply]
@stat0s2p I'm not sure what is incorrect about your setup, following the guide it worked correctly for me:
1. add breakpoints (added to both frontend and backend code (`browser` and `node`)
2. started a debug session with `launch browser backend`
3. started another debug session with `launch browser frontend` (this essentially will launch the application in a browser window).

It is possible you are experiencing the following bug in vscode (with their new preview debugger): https://github.com/microsoft/vscode/issues/102493.
A possible solution is to the stable debugger in the preferences:
```json
    "debug.javascript.usePreview": false
```
This worked correctly for me:

![image](https://user-images.githubusercontent.com/40359487/93092222-2b818c80-f66d-11ea-8bf4-16dccd9511f4.png)

**Reply by stat0s2p** [reply]
I add `"debug.javascript.usePreview": false` in my user config and finaly the debug breakpoint enabled. very thank you for your help.
![image](https://user-images.githubusercontent.com/24647459/93159834-ab792680-f741-11ea-81b9-606aae3c1a3e.png)
